#take seurat object turn it into single cell experiment
---
title: "cb_interaction.Rmd"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Exploring Seurat single cell objects 

In these scripts you'll be using data from our cerebellum study to demonstrate simple Seurat object manipulation and
visualizations (including ones similar to those in the original publication).

First download the data with gsutil (ask Aleks/Velina to give you read access to appropriate bucket). 
#```{bash, eval=FALSE, engine="sh"}
#gsutil cp gs://cerebellum_data/data/final_redo/seurat/rann/ubc_seurat2_3.RDS .
#```


```{r packages}
library(Seurat)
library(scater)
library(SingleCellExperiment)
library(ParetoTI)
```


```{r load_data}
# this is just my working directory -- you can skip this line
# setwd('/home/vkozarev/final_redo/seurat/rann/')
# ubc_s = readRDS('/Users/graceluettgen/BSSP_Pareto_cerebellum-master/ubc_seurat2_3.RDS')
ubc_s <- readRDS('ubc_seurat2_3.RDS')
ubc_s = UpdateSeuratObject(ubc_s)
ubc_sce <- as.SingleCellExperiment(ubc_s)
# these are some of the important metadata used
# including sex, cerebellar region (lobule), and named cluster
head(colData(ubc_sce))
# head(ubc_s@meta.data)
```

This object has a pre-computed tSNE embedding, which we can then use to visualize cluster labels, gene expression, and other features.

```{r plotting_ubcs}
# Plot with cluster labels
# ubc_s = UpdateSeuratObject(ubc_s)
# TSNEPlot(ubc_s, label = T)
 plotTSNE(ubc_sce, colour_by = "ident")

# Plot expression for some genes 
# FeaturePlot(ubc_s, c('Grm1', 'Calb2', 'Plcb4', 'Plcb1'), cols = c('yellow', 'purple'))

grm1 <- plotReducedDim(ubc_sce, 'TSNE', colour_by = "Grm1", point_alpha = 1, point_size = 2) + scale_colour_gradient2(low = "red", high = "purple", mid = "yellow", midpoint = 0, space = "Lab", na.value = "grey50", guide = "colourbar", aesthetics = c("colour", "fill")) + ggtitle("Grm1")

calb2 <- plotReducedDim(ubc_sce, 'TSNE', colour_by = "Calb2", point_alpha = 1, point_size = 2) + scale_colour_gradient2(low = "red", high = "purple", mid = "yellow", midpoint = 0, space = "Lab", na.value = "grey50", guide = "colourbar", aesthetics = c("colour", "fill")) + ggtitle("Calb2")

plcb4 <- plotReducedDim(ubc_sce, 'TSNE', colour_by = "Plcb4", point_alpha = 1, point_size = 2) + scale_colour_gradient2(low = "red", high = "purple", mid = "yellow", midpoint = 0, space = "Lab", na.value = "grey50", guide = "colourbar", aesthetics = c("colour", "fill")) + ggtitle("Plcb4")

plcb1 <- plotReducedDim(ubc_sce, 'TSNE', colour_by = "Plcb1", point_alpha = 1, point_size = 2) + scale_colour_gradient2(low = "red", high = "purple", mid = "yellow", midpoint = 0, space = "Lab", na.value = "grey50", guide = "colourbar", aesthetics = c("colour", "fill")) + ggtitle("Plcb1")

ggarrange(grm1, calb2, plcb4, plcb1)



# Plot and group by cerebellar region 
# DimPlot(ubc_s, reduction = 'tsne', group.by = 'region')
 plotReducedDim(ubc_sce, 'TSNE', colour_by = "region")
```

We can also easily calculate other types of lower dimensional embeddings, like UMAP. This object was converted to a Seurat object from a liger object, so carries 'inmf' data.
``` {r umap, echo=T, eval=FALSE}
#ubc_s = RunUMAP(ubc_s, reduction = 'inmf', dims = setdiff(1:15, 6))
ubc_sce = runUMAP(ubc_sce,   dimred="INMF", n_dimred=setdiff(1:15, 6), scale = T)

plotUMAP(ubc_sce, colour_by = "ident", ncomponents= c(2, 1))+ scale_y_reverse() + scale_x_reverse() + coord_flip()
#DimPlot(ubc_s, reduction = 'umap')
```

We can plot dot plots, which help us visualize the expresssion of markers across groups of cells.
``` {r dotplot}
# First log-normalized the data for better visualization
#ubc_s = NormalizeData(ubc_s)

# set the order for the levels
#ubc_s@active.ident = factor(ubc_s@active.ident, levels = c('Brinp2_On_UBCs', 'Intermediate_UBCs', 'Calb2_Off_UBCs'), ordered = T)

# test = logNormCounts(ubc_s.sce)
test = copy(ubc_sce)
# counts(test) = as.matrix(counts(test))
test = scran::computeSumFactors(test)
test = scater::logNormCounts(test, pseudo_count = 1)
# logcounts(test)[0:5, 0:5]
# test = normalizeData(test)
test$ident = factor(test$ident, levels = c('Brinp2_On_UBCs', 'Intermediate_UBCs', 'Calb2_Off_UBCs'), ordered = T)

DotPlot(ubc_s, features = c('Grm1', 'Plcb4', 'Calb2', 'Plcb1'))


plotDots(test, features = c('Grm1', 'Plcb4', 'Calb2', 'Plcb1'), group = "ident", low_color = "white", high_color = "blue") + labs(x = "Identity") + coord_flip() + theme(panel.grid.major = element_blank(), axis.line = element_line(colour = "black"))




```
